llms/mlx_lm/models/cohere.py

```diff
@@ -0,0 +1,200 @@
+from dataclasses import dataclass
+from typing import Optional, Tuple
+from functools import partial
+
+import mlx.core as mx
+import mlx.nn as nn
+
+from .base import BaseModelArgs
+
+
+@dataclass
+class ModelArgs(BaseModelArgs):
+    model_type: str
+    hidden_size: int = 8192
+    num_hidden_layers: int = 40
+    intermediate_size: int = 22528
+    num_attention_heads: int = 64
+    num_key_value_heads: int = 64
+    rope_theta: float = 8000000.0
+    vocab_size: int = 256000
+    tie_word_embeddings: bool = True
+    layer_norm_eps: float = 1e-05
+    logit_scale: float = 0.0625
+    attention_bias: bool = False
+
+
+@partial(mx.compile, shapeless=True)
+def ln_norm(x, eps, weight=None, bias=None):
+    t = x.dtype
+    x = x.astype(mx.float32)
+    means = mx.mean(x, axis=-1, keepdims=True)
+    var = mx.var(x, axis=-1, keepdims=True)
+    x = (x - means) * mx.rsqrt(var + eps)
+    x = x.astype(t)
+    h = weight.astype(mx.float32) * x
+    if bias is not None:
+        h = h + bias.astype(mx.float32)
+    return h
+
+class LayerNorm(nn.Module):
+    def __init__(self, dims: int, eps: float = 1e-5, bias=False):
+        super().__init__()
+        self.bias = mx.zeros((dims,)) if bias else None
+        self.weight = mx.ones((dims,))
+        self.eps = eps
+        self.dims = dims
+
+    def _extra_repr(self):
+        return f"{self.dims}, eps={self.eps}, affine={'weight' in self}"
+
+    def __call__(self, x: mx.array) -> mx.array:
+        if "weight" in self:
+            return ln_norm(x, self.eps, self.weight)
+        else:
+            return ln_norm(x, self.eps)
+
+
+class Attention(nn.Module):
+    def __init__(self, args: ModelArgs):
+        super().__init__()
+        self.args = args
+
+        dim = args.hidden_size
+        self.n_heads = n_heads = args.num_attention_heads
+        self.n_kv_heads = n_kv_heads = args.num_key_value_heads
+
+        head_dim = args.hidden_size // args.num_attention_heads
+        self.scale = head_dim**-0.5
+
+        attetion_bias = args.attention_bias
+
+        self.q_proj = nn.Linear(dim, n_heads * head_dim, bias=attetion_bias)
+        self.k_proj = nn.Linear(dim, n_kv_heads * head_dim, bias=attetion_bias)
+        self.v_proj = nn.Linear(dim, n_kv_heads * head_dim, bias=attetion_bias)
+        self.o_proj = nn.Linear(n_heads * head_dim, dim, bias=attetion_bias)
+
+        self.rope = nn.RoPE(head_dim, traditional=False, base=args.rope_theta)
```

Comment:
Make sure you pay attention to the RoPE implementation. It is the same as GPT-J and different from Llama :)

Reply:
Maybe I missed something, but it seems like they are using llama's rope in their model imp? https://huggingface.co/CohereForAI/c4ai-command-r-v01/blob/main/modeling_cohere.py#L95

Reply:
@saurabhdash thanks!

I'm following the original transformers implementation, the one @mzbac pointed. And as far as I can tell it's the same as llama implementation.

Please correct me if I'm missing something :)

Reply:
My bad I forgot to remove the comment! It is not Llama's implementation. The GPT-J version uses even and odd positions for rotation while the Llama one splits the tensor right in the middle.
https://github.com/huggingface/transformers/blob/main/src/transformers/models/gptj/modeling_gptj.py#L90
https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L184

If you use llama's implementation, it would seem like it's working but the generations will be bad.